### ISSUE TITLE
Component is not initialized, if it's internal

### ISSUE THREAD

**Issue by mkArtak** [dotnet/aspnetcore#21213]
I have a component defined as:

```razor
using System.Threading.Tasks;

namespace LittleBlazors.Components.Wizards
{
    internal class MarkerWizardPage : WizardPage
    {
        protected override async Task OnInitializedAsync()
        {
            base.OnInitialized();

            await this.Wizard.AllPagesLoaded();
        }
    }
}
```
I reference it in a different component markup as follows:

```razor
<MarkerWizardPage />
```

The `OnInitialized` never fires unless I change the visibility to `public` it does.

**Comment by javiercn** [comment]
@mkArtak components are required to be public classes. That's by design.

**Comment by mkArtak** [comment]
I get this, and I'm making a point there are situations, where you want them to be `internal`. Situations, where you want to have helper components in a component library, without exposing these helpers for outside usage.

**Comment by mkArtak** [comment]
I don't think there was similar ask, which means that nobody yet actually needed this. We'll probably need to backlog this and see whether the need for this becomes real or not.

**Comment by mrpmorris** [comment]
I've not needed it yet, but I am a strong advocate of using internal.

If the component can't be internal then it probably shouldn't compile at all rather than appearing to work but behaving unexpectedly.

**Comment by stefanloerwald** [comment]
I see internal components as quite useful. I have a few things that don't make sense for the user to see or even know about, thus internal would be a great way to hide them. The alternative is to stick them in another namespace, but that either requires me to separate the files (which is awkward) or use the `@namespace` directive, which I dislike (as it breaks the mental model of "namespace = RootNameSpace + folder hierarchy").

In short: `internal` components would help me declutter the set of components seen by the user.

**Comment by simonziegler** [comment]
I agree. They are as important as internal classes for a regular class library. I have a component library and have to prefix a set of components’ names with “Internal” in the hope that consumers will take the hint.

It would be a great enhancement to Blazor to allow internal components.

**Comment by mkArtakMSFT** [comment]
Closing this as a dupe of https://github.com/dotnet/aspnetcore/issues/5516
